aws/sdk/integration-tests/s3/tests/concurrency.rs

```diff
@@ -0,0 +1,379 @@
+            let mut semaphore_histogram_recorder = semaphore_histogram.recorder();
+            let mut req_histogram_recorder = req_histogram.recorder();
+
+            // because we wait on a permit from the semaphore, only <CONCURRENCY_LIMIT> futures
+            // will be run at once. Otherwise, we'd quickly get rate-limited by S3.
+            async move {
+                // Track time to acquire permit
+                let start = Instant::now();
+                let permit = semaphore
+                    .acquire()
+                    .await
+                    .expect("we'll get one if we wait long enough");
+                semaphore_histogram_recorder.saturating_record(start.elapsed().as_nanos() as u64);
+                debug!("semaphore_duration = {:?}", start.elapsed(),);
+                // Track time to send request and receive the response
+                let start = Instant::now();
+                let res = fut.await.expect("request should succeed");
+                debug!("put_object_duration = {:?}", start.elapsed(),);
+
+                req_histogram_recorder.saturating_record(start.elapsed().as_nanos() as u64);
+                drop(permit);
+                res
+            }
+            .instrument(debug_span!(REQUEST_SPAN_NAME, key))
+        })
+        .collect();
+
+    let task_creation_elapsed = task_creation_start.elapsed();
+
+    debug!(
+        "created {} PutObject futures in {task_creation_elapsed:?}",
+        TASK_COUNT
+    );
+    debug!("running futures concurrently with future::join_all");
+    let start = Instant::now();
+    let res: Vec<_> = future::join_all(futures).await;
+    debug!("future_joined_successfully, asserting that all tasks were run...");
+    assert_eq!(TASK_COUNT, res.len());
+    info!(
+        "all {TASK_COUNT} tasks completed after {:?}",
+        start.elapsed()
+    );
+
+    debug!("success! now, deleting test objects from S3...");
+    // Only 1,000 things can be deleted at once so the keys need to be divided into chunks
+    let chunked_keys: Vec<_> = object_keys.chunks(1000).collect();
+    for chunk in chunked_keys {
+        let mut delete_builder = Delete::builder();
+        for key in chunk {
+            delete_builder = delete_builder.objects(ObjectIdentifier::builder().key(key).build());
+        }
+
+        client
+            .delete_objects()
+            .bucket(BUCKET)
+            .delete(delete_builder.build())
+            .send()
+            .await
+            .unwrap();
+    }
+
+    info!("test data has been deleted from S3");
+
+    display_histogram(
+        "Semaphore Latency",
+        semaphore_histogram,
+        "s",
+        Duration::from_secs(1).as_nanos() as f64,
+    );
+    display_histogram(
+        "Request Latency",
+        req_histogram,
+        "s",
+        Duration::from_secs(1).as_nanos() as f64,
+    );
+}
+
+fn display_histogram(name: &str, mut h: SyncHistogram<u64>, unit: &str, scale: f64) {
```

Comment:
Not necessarily need to take an action, but since the histogram code is in place, it'd be cool to add a screenshot of example histograms to the PR if available.